### PR TITLE
[SPARK-37153][PYTHON] Inline type hints for python/pyspark/profiler.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for python/pyspark/profiler.py

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Let's get https://github.com/apache/spark/pull/34685 done first.

File: python/pyspark/profiler.py [hunk 1]
```diff
@@ -31,54 +36,51 @@ class ProfilerCollector(object):
...
+    def add_profiler(self, id: int, profiler: "Profiler") -> None:
         """Add a profiler for RDD `id`"""
         if not self.profilers:
             if self.profile_dump_path:
                 atexit.register(self.dump_profiles, self.profile_dump_path)
             else:
                 atexit.register(self.show_profiles)
 
-        self.profilers.append([id, profiler, False])
+        self.profilers.append([id, profiler, False])  # type: ignore[arg-type]
```
Comment (zero323) [inline python/pyspark/profiler.py:56]: Instead of ignoring this, it might make more sense to append `tuple`, but we'll have trace how it is used first.

File: python/pyspark/profiler.py [hunk 2]
```diff
@@ -31,54 +36,51 @@ class ProfilerCollector(object):
...
     """
 
-    def __init__(self, profiler_cls, dump_path=None):
-        self.profiler_cls = profiler_cls
-        self.profile_dump_path = dump_path
-        self.profilers = []
+    def __init__(self, profiler_cls: Type["Profiler"], dump_path: Optional[str] = None):
+        self.profiler_cls: Type[Profiler] = profiler_cls
+        self.profile_dump_path: Optional[str] = dump_path
+        self.profilers: List[Tuple[int, Profiler, bool]] = []
```
Comment (zero323) [inline python/pyspark/profiler.py:42]: It seems like we should change this to `List[List[Any]]`

Comment (ueshin) [reply python/pyspark/profiler.py:42]: May I ask why?

File: python/pyspark/profiler.py [hunk 3]
```diff
@@ -31,54 +36,51 @@ class ProfilerCollector(object):
...
-                self.profilers[i][2] = True
+                self.profilers[i][2] = True  # type: ignore[index]
 
 
 class Profiler(object):
     """
     PySpark supports custom profilers, this is to allow for different profilers to
     be used as well as outputting to different formats than what is provided in the
     BasicProfiler.
-
```
Comment (ueshin) [inline python/pyspark/profiler.py:73]: Could you revert these changes?

File: python/pyspark/profiler.py (hunk 2 above)
Comment (zero323) [reply python/pyspark/profiler.py]: @ueshin 

Well,it it seems it is a `list` of `lists` 

https://github.com/apache/spark/blob/e3256b838b5bd5c817bc95ba9d996b878078ad35/python/pyspark/profiler.py#L56

and it is used as such. In particular this part modifies records in place:

https://github.com/apache/spark/blob/e3256b838b5bd5c817bc95ba9d996b878078ad35/python/pyspark/profiler.py#L70

So we'd need type ignores in both places, just to cover for incorrect signature.

Having list of typed might be nice, but I am not sure if it acceptable change (do we depend on current mutable semantics, does some 3rd party code?) and it is definitely not something out of scope for this PR.

Comment (ueshin) [reply python/pyspark/profiler.py]: Got it. I think we can change it but not sure. Anyway we should do it in a separate PR if needed. Thanks.

**Review by ueshin — APPROVED** [review]
LGTM.

**Review by zero323 — COMMENTED** [review]
I'd like to give it another pass after we resolve SPARK-37083 / #34363.

Specifically, we should fill type hole for `PStatsParam` (`AccumulatorParam[Optional[pstats.Stats]]` with adjustment of `addInPlace`?) ,  but I'd like to have finalized version of `AccumulatorParam`.

**Comment by ueshin** [comment]
@dchvn Maybe could you rebase onto or merge master branch and see what happens?
@zero323 Now that #34363 has been merged. Could you give it another pass?

**Comment by zero323** [comment]
> @dchvn Maybe could you rebase onto or merge master branch and see what happens? @zero323 Now that #34363 has been merged. Could you give it another pass?

Sure thing @ueshin. I'll review it after it is synced with master and  https://github.com/apache/spark/pull/34731#pullrequestreview-819670002 is addressed.

**Comment by dchvn** [comment]
> > @dchvn Maybe could you rebase onto or merge master branch and see what happens? @zero323 Now that #34363 has been merged. Could you give it another pass?
> 
> Sure thing @ueshin. I'll review it after it is synced with master and [#34731 (review)](https://github.com/apache/spark/pull/34731#pullrequestreview-819670002) is addressed.

Updated, thanks!

File: python/pyspark/profiler.py [hunk 4]
```diff
@@ -142,11 +152,13 @@ class PStatsParam(AccumulatorParam):
     """PStatsParam is used to merge pstats.Stats"""
```
Comment (zero323) [inline python/pyspark/profiler.py:152]: This still has to be filled with full type, as per previous comments.

**Comment by itholic** [comment]
Looks pretty good to me except https://github.com/apache/spark/pull/34731#pullrequestreview-819670002

File: python/pyspark/profiler.py (hunk 4 above)
Comment (dchvn) [reply python/pyspark/profiler.py:152]: Sorry, I missed your previous comment! Updated :smiley:

**Review by zero323 — APPROVED** [review]

**Comment by zero323** [comment]
Merged into master.

Thanks everyone!

**Comment by dchvn** [comment]
Thanks all! ❤️
